Add TaskCompletionSource.SetCanceled(CancellationToken)

Frassle opened the pull request:
api-approved by #30862

Frassle committed:
Add TaskCompletionSource.SetCanceled(CancellationToken)

api-approved by #30862

Frassle committed:
Add to ref

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCompletionSource.cs:
```diff
@@ -337,5 +337,26 @@ public void SetCanceled()
+        /// <see cref="System.Threading.Tasks.TaskStatus.Canceled">Canceled</see>.
+        /// </exception>
+        /// <exception cref="System.ObjectDisposedException">The <see cref="Task"/> was disposed.</exception>
+        public void SetCanceled(CancellationToken cancellationToken)
```
The existing SetCanceled() can be changed now to just call SetCanceled(default), just as the equivalent TrySetCanceled delegates.

stephentoub commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -8624,6 +8624,7 @@ public partial class TaskCompletionSource<TResult>
         public TaskCompletionSource(System.Threading.Tasks.TaskCreationOptions creationOptions) { }
         public System.Threading.Tasks.Task<TResult> Task { get { throw null; } }
         public void SetCanceled() { }
+        public void SetCanceled(System.Threading.CancellationToken cancellationToken) { }
```
Please add tests as well.

Frassle committed:
SetCanceled(default)

Frassle committed:
Change some tests to use SetCaneled not TrySetCanceled

These tests used SetResult/SetException and TrySetCancelled so it could
pass a token. Changed to use SetCancelled to match the Result/Exception
useage.

stephentoub commented:
Thank you for tweaking some tests, but that's not really addressing my request. For example, there's no test I see validating that it throws the proper exception when the tcs is already completed. Can you please add a new test for that? I realize the test suite is a bit of a mess; it's never been rewritten / reorganized.

Frassle commented:
Planning on it, just haven't had time over the weekend yet to do so.

On Sun, 23 Feb 2020, 15:26 Stephen Toub, <notifications@github.com> wrote:

> Thank you for tweaking some tests, but that's not really addressing my
> request. For example, there's no test I see validating that it throws the
> proper exception when the tcs is already completed. Can you please add a
> new test for that? I realize the test suite is a bit of a mess; it's never
> been rewritten / reorganized.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/runtime/pull/32696?email_source=notifications&email_token=AAKVMAU2XAFINQJ5352Z6RLREKIRBA5CNFSM4KZRC5JKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMV6QJQ#issuecomment-590080038>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKVMATWF6I4TB6XGXDNDJDREKIRBANCNFSM4KZRC5JA>
> .
>

stephentoub commented:
Ah, no problem, I misinterpreted your commit that tweaked the tests as being complete. Thanks!

Frassle committed:
Add SetCanceled(CT) test

Frassle committed:
Check exception on re-cancel

Frassle committed:
Equal not Equals

Frassle committed:
Catch aggregate not TaskCancelled directly

Frassle committed:
Inner not exc

stephentoub commented on src/libraries/System.Threading.Tasks/tests/Task/TaskRtTests_Core.cs:
```diff
@@ -390,6 +390,49 @@ public static void RunTaskCompletionSourceTests_SetException()
+            catch
+            {
+                Assert.True(false, string.Format("RunTaskCompletionSourceTests:    > Error!  Canceled, but get-Result threw unexpected exception"));
+            }
```
This can just be:
```C#
AggregateException ae = Assert.Throws<AggregateException>(() => tcs.Task.Result);
TaskCanceledException tce = Assert.IsType<TaskCanceledException>(ae.InnerException);
Assert.Equal(cts.Token, tce.CancellationToken);
```

stephentoub commented on src/libraries/System.Threading.Tasks/tests/Task/TaskRtTests_Core.cs:
```diff
@@ -390,6 +390,49 @@ public static void RunTaskCompletionSourceTests_SetException()
+            {
+                Assert.True(false, string.Format("RunTaskCompletionSourceTests:    > Error!  Canceled, but re-Cancel threw unexpected exception"));
+
+            }
```
This can just be:
```C#
Assert.Throws<InvalidOperationException>(() => tcs.SetCanceled(cts.Token));
```

maryamariyan commented:
Note regarding the `new-api-needs-documentation` label:

This serves as a reminder for when your PR is modifying a ref *.cs file and adding/modifying public APIs, to please make sure the API implementation in the src *.cs file is documented with triple slash comments, so the PR reviewers can sign off that change.

Frassle committed:
Markup

Frassle committed:
s/m/n

stephentoub reviewed (approved):
Thank you for the PR!
